quartz/depgraph.ts

```diff
@@ -0,0 +1,189 @@
+export default class DepGraph<T> {
+  // node: incoming and outgoing edges
+  _graph = new Map<T, { incoming: Set<T>; outgoing: Set<T> }>()
+
+  constructor() {
+    this._graph = new Map()
+  }
+
+  export(): Object {
+    return {
+      nodes: this.nodes,
+      edges: this.edges,
+    }
+  }
+
+  toString(): string {
+    return JSON.stringify(this.export(), null, 2)
+  }
+
+  // BASIC GRAPH OPERATIONS
+
+  get nodes(): T[] {
+    return Array.from(this._graph.keys())
+  }
+
+  get edges(): [T, T][] {
+    let edges: [T, T][] = []
+    for (const [node, { outgoing }] of this._graph.entries()) {
+      outgoing.forEach((target) => {
+        edges.push([node, target])
+      })
+    }
+    return edges
+  }
+
+  hasNode(node: T): boolean {
+    return this._graph.has(node)
+  }
+
+  addNode(node: T): void {
+    if (!this._graph.has(node)) {
+      this._graph.set(node, { incoming: new Set(), outgoing: new Set() })
+    }
+  }
+
+  removeNode(node: T): void {
+    if (this._graph.has(node)) {
+      this._graph.delete(node)
+    }
+  }
+
+  hasEdge(from: T, to: T): boolean {
+    return this._graph.has(from) && this._graph.get(from)!.outgoing.has(to)
+  }
+
+  addEdge(from: T, to: T): void {
+    this.addNode(from)
+    this.addNode(to)
+
+    this._graph.get(from)!.outgoing.add(to)
+    this._graph.get(to)!.incoming.add(from)
+  }
+
+  removeEdge(from: T, to: T): void {
+    if (this._graph.has(from) && this._graph.has(to)) {
+      this._graph.get(from)!.outgoing.delete(to)
+      this._graph.get(to)!.incoming.delete(from)
+    }
+  }
+
+  outDegree(node: T): number {
+    return this._graph.get(node)!.outgoing.size
+  }
+
+  inDegree(node: T): number {
+    return this._graph.get(node)!.incoming.size
+  }
+
+  forEachOutNeighbor(node: T, callback: (neighbor: T) => void): void {
+    this._graph.get(node)!.outgoing.forEach(callback)
+  }
+
+  forEachInNeighbor(node: T, callback: (neighbor: T) => void): void {
+    this._graph.get(node)!.incoming.forEach(callback)
+  }
+
+  forEachEdge(callback: (edge: [T, T]) => void): void {
+    for (const [node, { outgoing }] of this._graph.entries()) {
+      outgoing.forEach((target) => {
+        callback([node, target])
+      })
+    }
+  }
+
+  // DEPENDENCY ALGORITHMS
+
+  // For the node provided:
+  // If node does not exist, add it
+  // If an edge was added in other, it is added in this graph
+  // If an edge was deleted in other, it is deleted in this graph
+  mergeEdgesForNode(other: DepGraph<T>, node: T): void {
```

Comment:
imo this needs a better name, maybe `updateEdgesFromDepGraph`

also weird that we only merges edges and not nodes (i guess we only update that one node but feels weird)

Reply:
We only update the edges for that one node. The idea being we change a file, the emitter will return a dep graph for that file, but it wont contain info about all the other files. So we only want to update the edges for that node.
Renamed to `updateIncomingEdgesForNode`